### PR TITLE
Dev/52 Integrate API [PATCH] /api/problems/{problemId} #52

### PR DESCRIPTION
**Pull request by chiyuyu**
- issue: #52 
 - componentize MarkdownEditor component
 
截圖 2021-04-29 下午1 40 44
截圖 2021-04-29 下午1 40 59
截圖 2021-04-29 下午1 41 17
截圖 2021-04-29 下午1 41 46

### CODE REVIEW

File: src/app/problem/edit/ProblemEditorTitle.css [hunk 1]
```diff
@@ -0,0 +1,51 @@
+.problem-editor .problem-editor-title {
+    margin-left: 150px;
+
+    font-family: Poppins, sans-serif;
+    font-style: normal;
+    font-weight: 600;
+    font-size: 35px;
```
Comment (Chaoyuuu) [inline src/app/problem/edit/ProblemEditorTitle.css:7]: use `scss` !!!

File: src/app/problem/edit/description/MarkdownEditor.js [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    // TODO: problemEditorService.getProblemDescription
+    const [editingState, setEditingState] = useState(false);
+    const [textareaVal, setTextareaVal] = useState('Press Edit Description to start writing the description. :smile: Styling with Markdown is supported. :+1:\n');
+    const [lastTextareaVal, setLastTextareaVal] = useState(textareaVal);
+
+    const handleDescription = e => {
+        e.preventDefault();
+        if (textareaVal === '') {
+            return;
+        }
```
Comment (Chaoyuuu) [inline src/app/problem/edit/description/MarkdownEditor.js:72]: Add `//TODO: empty information`.

File: src/app/problem/edit/description/MarkdownEditor.js [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+                </div>
+            </div>
+        </div>
+    );
+}
+
+function MarkdownEditor({problemId}) {
+    // TODO: problemEditorService.getProblemDescription
+    const [editingState, setEditingState] = useState(false);
+    const [textareaVal, setTextareaVal] = useState('Press Edit Description to start writing the description. :smile: Styling with Markdown is supported. :+1:\n');
```
Comment (Chaoyuuu) [inline src/app/problem/edit/description/MarkdownEditor.js:65]: remove `:smile:` and `:+1:`. We can't support it now. 😢

File: src/services/ProblemEditorService.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import axios from "axios";
+
+export class ProblemEditorService {
+
```
Comment (Chaoyuuu) [inline src/services/ProblemEditorService.js:4]: rename `ProblemEditorService` to `ProblemService`.

File: src/services/ProblemEditorService.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    constructor() {
+        this.axios = axios.create({
+            baseURL: process.env.REACT_APP_EXAM_SVC_BASE_URL,
+            timeout: 10000
```
Comment (Chaoyuuu) [inline src/services/ProblemEditorService.js:8]: Change baseURL to `process.env.REACT_APP_PROBLEM_SVC_BASE_URL`.
And add `REACT_APP_PROBLEM_SVC_BASE_URL=http://api.judgegirl.beta.pdlab.csie.ntu.edu.tw` in `.env` file.

File: src/app/problem/edit/description/MarkdownEditor.js [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+                    </ReactMarkdown>
+                </div>
+            </div>
+        </div>
+    );
+}
+
+function MarkdownEditor({problemId}) {
+    // TODO: problemEditorService.getProblemDescription
+    const [editingState, setEditingState] = useState(false);
```
Comment (Chaoyuuu) [inline src/app/problem/edit/description/MarkdownEditor.js:64]: Move `MarkdownEditor` to app/commons.

File: src/app/problem/edit/description/MarkdownEditor.js [hunk 7]
```diff
@@ -0,0 +1,146 @@
+import '../../ProblemEditor.css';
+import React, {useState} from "react";
```
Comment (Chaoyuuu) [inline src/app/problem/edit/description/MarkdownEditor.js:2]: Where did you use `ProblemEditor.css` ?
Don't use `ProblemEditor.css` when `MarkdownEditor` is a common component.

File: src/app/problem/edit/description/MarkdownEditor.js [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+            .then();
+    };
+
+    if (!editingState) {
+        return (
+            <div>
+                <div className="content-tabs">
+                    <div
+                        className={"content  active-markdown"}
+                    >
```
Comment (Chaoyuuu) [inline src/app/problem/edit/description/MarkdownEditor.js:86]: formate code.

File: src/app/problem/edit/description/Description.js [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+import React from "react";
+import './MarkdwonEditor.scss';
+import MarkdownEditor from "./MarkdownEditor";
+
+function Description({problemId}) {
+    return (
+        <div>
+            <SubtitleLine title={"Description"}/>
+            <MarkdownEditor problemId={problemId}/>
+        </div>
```
Comment (Chaoyuuu) [inline src/app/problem/edit/description/Description.js:12]: move it to `ProblemEditor`.

File: src/app/problem/edit/ProblemEditorTitle.js [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    const [problemNameInput, setProblemNameInput] = useState('Negative and Positive');
+    const [lastProblemNameInput, setLastProblemNameInput] = useState(problemNameInput);
+    const [editingState, setEditingState] = useState(false);
+
+    const handleSaveProblemName = (e) => {
+        e.preventDefault();
+        // prevent empty problem name
+        if (problemNameInput === '') {
+            return;
+        }
```
Comment (Chaoyuuu) [inline src/app/problem/edit/ProblemEditorTitle.js:19]: Add `//TODO: empty info`

File: src/app/problem/edit/ProblemEditorTitle.js [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+                        width={75} height={33}
+                        fontSize={15}
+                        borderRadius={20}
+                        borderColor={"#A2A3B1"}
+                        marginLeft={7}
+                        marginBottom={20}
+                        onClickFunc={() => {
+                            setEditingState(false);
+                            // Restore last saved problem name
+                            setProblemNameInput(lastProblemNameInput);
```
Comment (Chaoyuuu) [inline src/app/problem/edit/ProblemEditorTitle.js:61]: remove comment.

**Review by Chaoyuuu — CHANGES_REQUESTED** [review]
Review Done :)

File: src/app/commons/MarkdwonEditor.scss [hunk 12]
```diff
@@ -93,8 +89,13 @@
 
   .active-markdown {
     display: block;
-    padding: 15px;
     overflow: scroll;
+    background: #FFFFFF;
+    border: 1px solid #E4E1E1;
+    box-sizing: border-box;
+    border-radius: 10px;
+    width: 756px;
```
Comment (Chaoyuuu) [inline src/app/commons/MarkdwonEditor.scss:97]: remember to extract common CSS !!
e.g., `background: #FFFFFF`, `display: block`, `box-sizing: border-box`, etc.

**Review by Chaoyuuu — CHANGES_REQUESTED** [review]
Review Done !

File: src/app/commons/MarkdwonEditor.scss [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+      border: 1px solid #C6C4C4;
+      border-bottom: transparent;
+      border-radius: 10px 10px 0px 0px;
+      position: relative;
+      top: 1px;
+    }
+
+    .tab-write {
+      /* Write */
+      cursor: pointer;
```
Comment (Chaoyuuu) [inline src/app/commons/MarkdwonEditor.scss:33]: remove comment :)

File: src/app/commons/MarkdwonEditor.scss [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+      width: 44px;
+      height: 26px;
+      left: 0px;
+      top: calc(50% - 26px / 2);
+      margin: 7px 16px 5px 18px;
+    }
+
+    .tab-preview {
+      /* Preview */
+      cursor: pointer;
```
Comment (Chaoyuuu) [inline src/app/commons/MarkdwonEditor.scss:44]: ditto

File: src/services/ProblemService.js [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+        this.axios = axios.create({
+            baseURL: process.env.REACT_APP_PROBLEM_SVC_BASE_URL,
+            timeout: 10000
+        });
+    }
+
+    async modifyProblemTitle(problemId, title) {
+        return this.axios.patch(`/api/problems/${problemId}`, {problemId, title})
+            .then()
+    }
```
Comment (Chaoyuuu) [inline src/services/ProblemService.js:15]: redundant, remove this.

remember to invoke `.then()` on the client-side

File: src/services/ProblemService.js [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+
+    async modifyProblemDescription(problemId, description) {
+        return this.axios.patch(`/api/problems/${problemId}`, {problemId, description})
+            .then();
+    }
```
Comment (Chaoyuuu) [inline src/services/ProblemService.js:20]: ditto

File: src/app/commons/MarkdwonEditor.scss [hunk 17]
```diff
@@ -0,0 +1,80 @@
+.MarkdownEditor {
+  background: #FFFFFF;
+  display: block;
```
Comment (Chaoyuuu) [inline src/app/commons/MarkdwonEditor.scss:3]: use variable.
```css
$markdown-background-color: #FFFFFF;

.MarkdownEditor {
  background: $markdown-background-color;
  display: block;
  ...
}

File: src/app/commons/MarkdwonEditor.scss [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+    font-size: 17px;
+    line-height: 25px;
+    display: flex;
+    align-items: center;
+    color: #6D6E7D;
+
+    .active-tabs {
+      // to hide border-button line
+      background: white;
+      border: 1px solid #C6C4C4;
```
Comment (Chaoyuuu) [inline src/app/commons/MarkdwonEditor.scss:24]: `background: $markdown-background-color;`

**Review by Chaoyuuu — CHANGES_REQUESTED** [review]
Last things to change 👍🏻

**Review by Chaoyuuu — APPROVED** [review]
LGTM

File: src/app/commons/MarkdwonEditor.scss [hunk 19]
```diff
@@ -0,0 +1,80 @@
+$markdown-background-color: #FFFFFF;
+
+.MarkdownEditor {
```
Comment (Johnny850807) [inline src/app/commons/MarkdwonEditor.scss:3]: rename to `markdown-editor`

File: src/app/problem/edit/ProblemEditorTitle.js [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+function ProblemEditorTitle ({problemId}) {
+    // TODO: problemService.getProblemTitle
+    const [problemNameInput, setProblemNameInput] = useState('Negative and Positive');
+    const [lastProblemNameInput, setLastProblemNameInput] = useState(problemNameInput);
+    const [editingState, setEditingState] = useState(false);
+
+    const handleSaveProblemName = (e) => {
+        e.preventDefault();
+        // TODO: empty problem name notification
+        if (problemNameInput === '') {
```
Comment (Johnny850807) [inline src/app/problem/edit/ProblemEditorTitle.js:16]: `problemNameInput.isEmpty()`

File: src/app/problem/edit/ProblemEditorTitle.js [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+            return;
+        }
+
+        setEditingState(false);
+
+        problemService.modifyProblemTitle(problemId, problemNameInput)
+            .then();
```
Comment (Johnny850807) [inline src/app/problem/edit/ProblemEditorTitle.js:23]: `.then(() => console.log('The problem's title has been modified'))`

**Review by Johnny850807 — CHANGES_REQUESTED** [review]
Review done

**Review by Johnny850807 — APPROVED** [review]
ＬＧＴＭ

**Pull request merged**